### PR TITLE
Move caching to node executor for fast cache hits

### PR DESCRIPTION
**Pull request by hamersaw**
# TL;DR
Currently cache lookups require 3 node-level phase changes, which equates to 3 streak rounds of flytepropeller execution. The overhead of updating the k8s FlyteWorkflow CRD after each of these rounds may make it seem like cache lookups are slow. This PR updates cache lookups to happen in the node executor, which allows us to transition nodes directly from NotYetStarted to Success and immediately process downstream nodes in the same FlytePropeller round making cache lookup overhead minimal.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
In addition to the performance improvements this PR refactors the code to support future implementation of caching other Flyte node types (ie. subworkflow / launchplan and array nodes when they are implemented).

## Tracking Issue
fixes https://github.com/flyteorg/flyte/issues/2886

## Follow-up issue
_NA_

### CODE REVIEW

File: pkg/controller/nodes/array/handler.go [hunk 1]
```diff
@@ -295,6 +295,18 @@ func (a *arrayNodeHandler) Handle(ctx context.Context, nCtx interfaces.NodeExecu
 
 			retryAttempt := subNodeStatus.GetAttempts()
 
+			// fastcache will not emit task events for cache hits. we need to manually detect a
+			// transition to `SUCCEEDED` and add an `ExternalResourceInfo` for it.
+			if cacheStatus == idlcore.CatalogCacheStatus_CACHE_HIT && len(arrayEventRecorder.TaskEvents()) == 0 {
```
Comment (kumare3) [inline pkg/controller/nodes/array/handler.go:300]: is there one potential problem here, consider this case

Round 1:
  node 1 - cache-hit
   node 2 - cache-hit
   node 3 - system error
Round 2:
  node 1 - cache hit
  node 2 - cache miss
      - event write failure?

File: pkg/controller/nodes/array/handler.go [hunk 2]
```diff
@@ -476,6 +488,7 @@ func (a *arrayNodeHandler) Handle(ctx context.Context, nCtx interfaces.NodeExecu
 			arrayNodeState.TaskPhaseVersion = taskPhaseVersion + 1
 		}
 
+		fmt.Printf("HAMERSAW - sending event with taskPhase %+v and version %d\n", taskPhase, taskPhaseVersion)
```
Comment (kumare3) [inline pkg/controller/nodes/array/handler.go:491]: remove this?

File: pkg/controller/nodes/array/handler.go (hunk 1 above)
Comment (kumare3) [reply pkg/controller/nodes/array/handler.go:300]: Or maybe this will not happen, as when we actually get a system error I think we update the state?
is it possible to simulate this and test this?

One other option is allow admin to move the state from success to running!

**Review by kumare3 — COMMENTED** [review]
I have one major comment, otherwise the code looks so much better.
This makes it actually way more readable. Not sure why the caching stuff was at the task level at all 🗡️

File: pkg/controller/nodes/array/handler.go (hunk 2 above)
Comment (hamersaw) [reply pkg/controller/nodes/array/handler.go:491]: doh

**Review by kumare3 — APPROVED** [review]
thank you for the explanation

**Review by kumare3 — APPROVED** [review]
